### PR TITLE
Environment section [1 of n] (Multiple Items)

### PR DESCRIPTION
**Pull request by Poltergeist**
env_vars mov

This is also *Renaming* Duplicatable Rows to Multiple Items, as this is a more fitting name.

This is adding the support for multiple Items.

As in #1317 You need to add some changes to get this working.

Since most of this has been achieved during the task of the Multiple Items(Duplicatable Rows) integration, however this is not the complete environment section but does contain some parts.

TODO:
- [x] Use Parsers on JSON change events
- [x] Tidy up reducers JSON & Form
- [x] Add Style

### CODE REVIEW

**Comment by orlandohohmeier** [comment]
THIS IS AMAZING! ✨

**Comment by jfurrow** [comment]
Wow! This looks awesome so far!

File: plugins/services/src/js/components/forms/EnvironmentFormSection.js [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+  }
+
+  getEnvironmentLines(data) {
+    const {errors} = this.props;
+
+    return data.map((env, key) => {
+      let keyLabel = null;
+      let valueLabel = null;
+      if (key === 0) {
+        keyLabel = (
```
Comment (mlunoe) [inline plugins/services/src/js/components/forms/EnvironmentFormSection.js]: We can do this inline: `keyLabel = <FieldLabel>Key</FieldLabel>;`

File: plugins/services/src/js/components/forms/EnvironmentFormSection.js [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+            required={false}
+            showError={Boolean(errors.env[key])}>
+            {keyLabel}
+            <FieldInput
+              name={`env.${key}.key`}
+              type="text"
+              value={env.key}/>
+            <FieldError>{errors.env[key]}</FieldError>
+            <span className="emphasis"
+              style={{
```
Comment (mlunoe) [inline plugins/services/src/js/components/forms/EnvironmentFormSection.js]: Let's not use inline styles like this

File: plugins/services/src/js/components/forms/EnvironmentFormSection.js [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+    ); });
+  }
+
+  getLabelsLines(data) {
+    const {errors} = this.props;
+    return data.map((label, key) => {
+      let keyLabel = null;
+      let valueLabel = null;
+      if (key === 0) {
+        keyLabel = (
```
Comment (mlunoe) [inline plugins/services/src/js/components/forms/EnvironmentFormSection.js]: Let's do this inline

File: plugins/services/src/js/components/forms/EnvironmentFormSection.js [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+            required={false}
+            showError={Boolean(errors.labels[key])}>
+            {keyLabel}
+            <FieldInput
+              name={`labels.${key}.key`}
+              type="text"
+              value={label.key}/>
+            <span className="emphasis"
+              style={{
+                position: 'absolute',
```
Comment (mlunoe) [inline plugins/services/src/js/components/forms/EnvironmentFormSection.js]: No inline styles, please

File: plugins/services/src/js/components/forms/EnvironmentFormSection.js [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+            {valueLabel}
+            <FieldInput
+              name={`labels.${key}.value`}
+              type="text"
+              value={label.value}/>
+            <FieldError>{errors.labels[key]}</FieldError>
+          </FormGroup>
+          <div className="form-group flex flex-item-align-end column-4">
+            <a className="column-3 button button-primary-link button-flush"
+              onClick={(event) => {
```
Comment (mlunoe) [inline plugins/services/src/js/components/forms/EnvironmentFormSection.js]: Let's create a `handleRemoveItem` function to separate view declaration and logic. Also, so we are not passing down a new function on each render

File: plugins/services/src/js/components/forms/EnvironmentFormSection.js [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+            Environment Variables
+          </h2>
+          <p>
+            Set up environment variables for each task your service launches.
+          </p>
+        </div>
+        {this.getEnvironmentLines(data.env)}
+        <div>
+          <a className="button button-primary-link button-flush"
+            onClick={(event) => {
```
Comment (mlunoe) [inline plugins/services/src/js/components/forms/EnvironmentFormSection.js]: Let's create a handler

File: plugins/services/src/js/components/forms/EnvironmentFormSection.js [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+          </h2>
+          <p>
+            Attach metadata to expose additional information to other services.
+          </p>
+        </div>
+        {this.getLabelsLines(data.labels)}
+        <div>
+          <a
+            className="button button-primary-link button-flush"
+            onClick={(event) => {
```
Comment (mlunoe) [inline plugins/services/src/js/components/forms/EnvironmentFormSection.js]: Let's create a handler

File: plugins/services/src/js/reducers/form/EnvironmentVariables.js [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+
+        return this.env.reduce((memo, item) => {
+          if (item.key != null || item.value != null) {
+            memo[item.key] = item.value;
+          }
+          return memo;
+        }, {});
+      }
+
+      const joinedPath = path.join('.');
```
Comment (mlunoe) [inline plugins/services/src/js/reducers/form/EnvironmentVariables.js]: Let' move this up. I see three `path.join('.')`s now.
You could do `const joinedPath = Array.isArray(path) && path.join('.');` (or something similar)

File: plugins/services/src/js/reducers/form/EnvironmentVariables.js [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+  ADD_ITEM,
+  REMOVE_ITEM,
+  SET
+} from '../../../../../../src/js/constants/TransactionTypes';
+import Transaction from '../../../../../../src/js/structs/Transaction';
+
+module.exports = {
+  JSONReducer(state, {type, path, value}) {
+    if (this.env == null) {
+      this.env = [];
```
Comment (mlunoe) [inline plugins/services/src/js/reducers/form/EnvironmentVariables.js]: Not exactly sure what is going on here, what is this `env` array? Could you maybe add some comments? This is hard to decipher.

File: plugins/services/src/js/reducers/form/EnvironmentVariables.js [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+      return memo;
+    }, {});
+  },
+  JSONParser(state) {
+    if (state.env == null) {
+      return [];
+    }
+
+    return Object.keys(state.env).reduce(function (memo, key, index) {
+      memo.push(new Transaction(['env'], index, ADD_ITEM));
```
Comment (mlunoe) [inline plugins/services/src/js/reducers/form/EnvironmentVariables.js]: Why is this necessary?

File: plugins/services/src/js/reducers/form/EnvironmentVariables.js [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+          case REMOVE_ITEM:
+            state = state.filter((item, index) => {
+              return index !== value;
+            });
+            break;
+        }
+        return state;
+      }
+
+      let joinedPath = path.join('.');
```
Comment (mlunoe) [inline plugins/services/src/js/reducers/form/EnvironmentVariables.js]: Let's do the same with this joinedPath

File: plugins/services/src/js/reducers/form/EnvironmentVariables.js [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+
+      return memo;
+    }, []);
+  },
+  FormReducer(state = [], {type, path, value}) {
+    if (path != null && path.join('.').search('env') !== -1) {
+      if (path.join('.') === 'env') {
+        switch (type) {
+          case ADD_ITEM:
+            state.push({key: null, value: null});
```
Comment (mlunoe) [inline plugins/services/src/js/reducers/form/EnvironmentVariables.js]: Are we ok with modifying `state` here? Maybe we should return a new object?

File: plugins/services/src/js/reducers/form/Labels.js [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+  ADD_ITEM,
+  REMOVE_ITEM,
+  SET
+} from '../../../../../../src/js/constants/TransactionTypes';
+import Transaction from '../../../../../../src/js/structs/Transaction';
+
+module.exports = {
+  JSONReducer(state, {type, path, value}) {
+    if (this.labels == null) {
+      this.labels = [];
```
Comment (mlunoe) [inline plugins/services/src/js/reducers/form/Labels.js]: Let's get a comment about the purpose of this variable

File: plugins/services/src/js/reducers/form/Labels.js [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+      memo[item.key] = item.value;
+      return memo;
+    }, {});
+  },
+  JSONParser(state) {
+    if (state.labels == null) {
+      return [];
+    }
+    return Object.keys(state.labels).reduce(function (memo, key, index) {
+      memo.push(new Transaction(['labels'], index, ADD_ITEM));
```
Comment (mlunoe) [inline plugins/services/src/js/reducers/form/Labels.js]: I'm getting confused about this again

File: src/js/components/form/FormGroup.js [hunk 15]
```diff
@@ -8,7 +8,7 @@ const FormGroup = (props) => {
   let {children, className, errorClassName, showError, required} = props;
 
   let clonedChildren = React.Children.map(children, (child) => {
-    if (!showError && child.type === FieldError) {
+    if (child == null || (!showError && child.type === FieldError)) {
```
Comment (mlunoe) [inline src/js/components/form/FormGroup.js]: 👍

File: plugins/services/src/js/components/forms/EnvironmentFormSection.js (hunk 1 above)
Comment (Poltergeist) [inline plugins/services/src/js/components/forms/EnvironmentFormSection.js]: done

File: plugins/services/src/js/components/forms/EnvironmentFormSection.js (hunk 3 above)
Comment (Poltergeist) [inline plugins/services/src/js/components/forms/EnvironmentFormSection.js]: done

File: plugins/services/src/js/components/forms/EnvironmentFormSection.js (hunk 5 above)
Comment (Poltergeist) [inline plugins/services/src/js/components/forms/EnvironmentFormSection.js]: Done.

File: plugins/services/src/js/components/forms/EnvironmentFormSection.js (hunk 6 above)
Comment (Poltergeist) [inline plugins/services/src/js/components/forms/EnvironmentFormSection.js]: done

File: plugins/services/src/js/components/forms/EnvironmentFormSection.js (hunk 7 above)
Comment (Poltergeist) [inline plugins/services/src/js/components/forms/EnvironmentFormSection.js]: done

File: plugins/services/src/js/reducers/form/EnvironmentVariables.js (hunk 9 above)
Comment (Poltergeist) [inline plugins/services/src/js/reducers/form/EnvironmentVariables.js]: added a comment, which explains what this is referring to and why it is used here.

File: plugins/services/src/js/reducers/form/EnvironmentVariables.js (hunk 10 above)
Comment (Poltergeist) [inline plugins/services/src/js/reducers/form/EnvironmentVariables.js]: This is necessary as the Reducer is adding a object with `{key: null, value: null}` to the array. If this would not happen we could run into problems, like an Item is not an Object. Which we have to check on each render. I would rather be more verbose here but remove therefore complexity from views and reducers.

File: plugins/services/src/js/reducers/form/EnvironmentVariables.js (hunk 12 above)
Comment (Poltergeist) [inline plugins/services/src/js/reducers/form/EnvironmentVariables.js]: We should be okay. As the input state from the combine reducers ist already a copy https://github.com/dcos/dcos-ui/blob/master/src/js/utils/ReducerUtil.js#L39

File: plugins/services/src/js/reducers/form/EnvironmentVariables.js (hunk 11 above)
Comment (Poltergeist) [inline plugins/services/src/js/reducers/form/EnvironmentVariables.js]: check

File: plugins/services/src/js/reducers/form/EnvironmentVariables.js (hunk 10 above)
Comment (Poltergeist) [inline plugins/services/src/js/reducers/form/EnvironmentVariables.js]: Also we need this to add a item / row in the form. Otherwise we would have logic in the form.  which should be in the Batch. But I am open to another solution here.

File: plugins/services/src/js/components/forms/EnvironmentFormSection.js [hunk 16]
```diff
@@ -0,0 +1,213 @@
...
+    const {errors} = this.props;
+    return data.map((label, key) => {
+      let keyLabel = null;
+      let valueLabel = null;
+      if (key === 0) {
+        keyLabel = <FieldLabel>Key</FieldLabel>;
+        valueLabel = <FieldLabel>Value</FieldLabel>;
+      }
+      return (
+        <div key={key} className="flex row">
```
Comment (jfurrow) [inline plugins/services/src/js/components/forms/EnvironmentFormSection.js]: @mlunoe @Poltergeist Looks like we should create a `<FormRow>` component to do this for us; what do you think? (no need to do it in this PR)

File: plugins/services/src/js/components/forms/EnvironmentFormSection.js [hunk 17]
```diff
@@ -0,0 +1,213 @@
...
+            required={false}
+            showError={Boolean(errors.env[key])}>
+            {valueLabel}
+            <FieldInput
+              name={`env.${key}.value`}
+              type="text"
+              value={env.value}/>
+            <FieldError>{errors.env[key]}</FieldError>
+          </FormGroup>
+          <div className="form-group flex flex-item-align-end column-4">
```
Comment (jfurrow) [inline plugins/services/src/js/components/forms/EnvironmentFormSection.js]: Let's use the `FormGroup` component here.

File: plugins/services/src/js/components/forms/EnvironmentFormSection.js (hunk 16 above)
Comment (mlunoe) [inline plugins/services/src/js/components/forms/EnvironmentFormSection.js]: But how would the columns then work? We had issues with this in the old form. It is a good conversation to have, but let's take somewhere else?

File: plugins/services/src/js/components/forms/EnvironmentFormSection.js [hunk 18]
```diff
@@ -0,0 +1,213 @@
...
+            <a className="column-3 button button-primary-link button-flush"
```
Comment (jfurrow) [inline plugins/services/src/js/components/forms/EnvironmentFormSection.js]: Let's remove the `column-3` class here. `button-flush` isn't removing the padding because we're on an outdated version of CNVS, but once that's up to date, the button should be aligned nicely :)

File: plugins/services/src/js/components/forms/EnvironmentFormSection.js [hunk 19]
```diff
@@ -0,0 +1,213 @@
...
+              data-path="env"
```
Comment (jfurrow) [inline plugins/services/src/js/components/forms/EnvironmentFormSection.js]: I would much prefer us a more React friendly approach rather than the element's `dataset`. I would suggest making a `MultipleItemDelete` component (probably that's not a good name), to which we can pass these `path` and `value` and attach click handlers.

File: plugins/services/src/js/components/forms/EnvironmentFormSection.js [hunk 20]
```diff
@@ -0,0 +1,213 @@
...
+          </p>
+        </div>
+        {this.getEnvironmentLines(data.env)}
+        <div>
+          <a className="button button-primary-link button-flush"
+            data-value={data.env.length}
+            data-path="env"
+            onClick={this.handleOnAddItem}>
+            + Add Environment Variable
+          </a>
```
Comment (jfurrow) [inline plugins/services/src/js/components/forms/EnvironmentFormSection.js]: I would like for us not to use the `dataset` for this add button, similar to my comment about the delete button.

**Comment by jfurrow** [comment]
@Poltergeist This works REALLY nicely! I did request a couple of changes.

Also, what do you think about trying to restrict users from deleting items when it is the last of the multiple items? So, if there are two or more multiple items, they can delete any item they wish. But as soon as there is only one multiple item, they cannot delete it.

Here's a terrible gif!

File: plugins/services/src/js/components/forms/EnvironmentFormSection.js (hunk 20 above)
Comment (mlunoe) [inline plugins/services/src/js/components/forms/EnvironmentFormSection.js]: done

File: plugins/services/src/js/components/forms/EnvironmentFormSection.js (hunk 19 above)
Comment (mlunoe) [inline plugins/services/src/js/components/forms/EnvironmentFormSection.js]: done

File: plugins/services/src/js/components/forms/EnvironmentFormSection.js (hunk 18 above)
Comment (mlunoe) [inline plugins/services/src/js/components/forms/EnvironmentFormSection.js]: done
